internal/asm/buffer.go

```diff
@@ -0,0 +1,188 @@
+package asm
+
+import (
+	"encoding/binary"
+	"fmt"
+	"unsafe"
+
+	"github.com/tetratelabs/wazero/internal/platform"
+)
+
+var zero [16]byte
+
+// CodeSegment represents a memory mapped segment where native CPU instructions
+// are written.
+//
+// To construct code segments, the program must call Next to obtain a buffer
+// view capable of writing data at the end of the segment. Next must be called
+// before generating the code of a function because it aligns the next write on
+// 16 bytes.
+//
+// Instances of CodeSegment hold references to memory which is NOT managed by
+// the garbage collector and therefore must be released *manually* by calling
+// their Unmap method to prevent memory leaks.
+type CodeSegment struct {
+	code []byte
+	size int
+}
+
+func MakeCodeSegment(code []byte) CodeSegment {
+	return CodeSegment{code: code}
+}
+
+func (seg *CodeSegment) Map(size int) error {
+	if seg.code != nil {
+		return fmt.Errorf("code segment already initialized to memory mapping of size %d", len(seg.code))
+	}
+	b, err := platform.MmapCodeSegment(size)
+	if err != nil {
+		return err
+	}
+	seg.code = b
+	seg.size = 0
+	return nil
+}
+
+func (seg *CodeSegment) Unmap() error {
```

Comment:
If Map is called multiple times, I might suggest renaming this to Close instead.

Reply:
I chose `Unmap` because the value may still be used after unmapping it, I felt like `Close` implied that the value wouldn't be usable anymore (e.g. like closing a file or a stream).

Programs also tend to conventionally ignore errors returned by `Close() error`, I wanted to trigger the reader's attention around how they should handle errors returned when unmapping code segments.

I'll enhance the documentation but let me know if you feel like naming it `Close` has upsides that I'm missing out on.